File: src/main/java/net/exotia/exotiaprofanity/listeners/PlayerChatListener.java
package net.exotia.exotiaprofanity.listeners;

import net.exotia.exotiaprofanity.Permissions;
import net.exotia.exotiaprofanity.client.ProfanityClient;
import net.exotia.exotiaprofanity.configuration.Configuration;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;

public class PlayerChatListener implements Listener {
    private final ProfanityClient profanityClient;
    private final Configuration configuration;

    public PlayerChatListener(ProfanityClient profanityClient) {
        this.profanityClient = profanityClient;
        this.configuration = profanityClient.getConfiguration();
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onChat(AsyncPlayerChatEvent event) {
        Player player = event.getPlayer();
        if (player.hasPermission(Permissions.PROFANITY_BYPASS)) return;
        if (!event.isAsynchronous()) return;
        if (!this.profanityClient.isOpen()) this.profanityClient.connect();
        this.profanityClient.send("/censor " + this.format(event.getMessage(), player));
        event.setCancelled(true);
    }

    private String format(String playerMessage, Player player) {
        String delimiter = this.configuration.getDelimiter();
        return String.join(delimiter, player.getUniqueId().toString(), playerMessage.replace(delimiter, " "));
    }
}

File: src/main/java/net/exotia/exotiaprofanitykotlin/listeners/PlayerChatListener.kt
package net.exotia.exotiaprofanitykotlin.listeners

import net.exotia.exotiaprofanitykotlin.configuration.Configuration
import okhttp3.WebSocket
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import java.util.logging.Logger

class PlayerChatListener(
  private var ws: WebSocket,
  private val logger: Logger,
  private val configuration: Configuration,
) : Listener {
  private val profanityBypass = "exotia.profanity.bypass"

  @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
  fun onChat(event: AsyncPlayerChatEvent) {
    val player: Player = event.player
    if (player.hasPermission(profanityBypass)) return
    if (!event.isAsynchronous) return
    val wsResponse = this.ws.send("/censor ${this.format(event.message, player)}")
    if (wsResponse == true) {
      event.isCancelled = true
    }
  }

  private fun format(playerMessage: String, player: Player): String {
    val delimiter: String = this.configuration.delimiter
    return player.uniqueId.toString() + delimiter + playerMessage.replace(delimiter, " ")
  }
}
